crates/dt-workspace/src/lib.rs

```diff
@@ -1,54 +1,124 @@
-#[derive(Debug, Error)]
-pub enum WorkspaceError {
-    #[error("Failed to load config")]
-    Config(#[from] ConfigError),
+/// The result given by [`Workspace::find_workspace_dir`].
+///
+/// Ordering of heuristics:
+///
+/// 1. `TomlConfig`
+/// 2. `LinuxMarker`
+/// 3. `Fallback`
+#[derive(Debug)]
+pub enum WorkspacePathFindResult<'start> {
+    /// TOML config file ([`toml_config::CONFIG_FILENAME`]) is a sibling of the start directory or its ancestors.
+    TomlConfig {
+        toml_file_path: PathBuf,
+        workspace_dir: &'start Path,
+    },
+    /// Linux kernel (Using the marker file [`LINUX_MARKER`]) is a sibling of the start directory or its ancestors.
+    LinuxMarker { workspace_dir: &'start Path },
+    /// No workspace directory found using the other heuristics. Use the start directory.
+    Fallback { workspace_dir: &'start Path },
+}
+
+impl<'start> WorkspacePathFindResult<'start> {
+    #[must_use]
+    pub fn toml_file_path(&self) -> Option<&PathBuf> {
+        match self {
+            Self::TomlConfig {
+                toml_file_path,
+                workspace_dir: _,
+            } => Some(toml_file_path),
+            _ => None,
+        }
+    }
+
+    #[must_use]
+    pub fn workspace_dir(&self) -> &'start Path {
+        match self {
+            Self::TomlConfig {
+                toml_file_path: _,
+                workspace_dir,
+            }
+            | Self::LinuxMarker { workspace_dir }
+            | Self::Fallback { workspace_dir } => workspace_dir,
+        }
+    }
 }
 
 #[cfg(test)]
 mod tests {
+    use std::env::current_dir;
+
     use super::*;
 
-    #[test]
-    fn load() {
-        let workspace = Workspace::try_new(PathBuf::from("test_data")).unwrap();
-        assert_eq!(vec![PathBuf::from(".")], workspace.config.include_paths);
+    impl<'start> WorkspacePathFindResult<'start> {
+        pub fn unwrap_toml(self) -> (PathBuf, &'start Path) {
+            match self {
+                Self::TomlConfig {
+                    toml_file_path,
+                    workspace_dir,
+                } => (toml_file_path, workspace_dir),
+                _ => panic!("not a TomlConfig"),
+            }
+        }
+    }
+
+    fn cwd_with<const N: usize>(join: [&str; N]) -> PathBuf {
+        join.iter()
+            .fold(current_dir().unwrap(), |p, j| p.join(j))
+            .canonicalize()
+            .unwrap()
     }
 
     #[test]
-    fn load_ancestor() {
-        // this is equal to `Workspace::try_new(PathBuf::from("not_exists"))`
-        // because we'll fallback to ancestor config
-        let workspace = Workspace::try_new(PathBuf::from(".")).unwrap();
+    fn find() {
+        // toml
+        let target = current_dir().unwrap();
         assert_eq!(
-            vec![PathBuf::from("test_data")],
-            workspace.config.include_paths
+            (cwd_with([toml_config::CONFIG_FILENAME]), target.as_path()),
+            Workspace::find_workspace_dir(target.as_path()).unwrap_toml()
```

Comment:
Actual result before, expected result after.
```suggestion
            Workspace::find_workspace_dir(target.as_path()).unwrap_toml(),
            (cwd_with([toml_config::CONFIG_FILENAME]), target.as_path())
```

Why is there a `cwd_with`? You can just do `Path::new(CONFIG_FILENAME).canonicalize().unwrap()`, or `Path::new(".").join(CONFIG_FILENAME).canonicalize().unwrap()` (basically does the same thing but may look cleaner).

`canonicalize` and `current_dir` are mutually exclusive. You only need one or the other.

Reply:
Should the actual result be first? Although dt-tools using this convention, it requires polluting code with unnecessary generics:
```
error[E0283]: type annotations needed
   --> crates/dt-workspace/src/config/env_config.rs:25:13
    |
25  |             envy::from_iter([("INCLUDE_PATHS".into(), "linux,linux-old".into())]).unwrap(),
    |             ^^^^^^^^^^^^^^^ cannot infer type of the type parameter `T` declared on the function `from_iter`
    |
    = note: cannot satisfy `_: DeserializeOwned`
note: required by a bound in `envy::from_iter`
   --> /home/roma/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/envy-0.4.2/src/lib.rs:303:8
    |
301 | pub fn from_iter<Iter, T>(iter: Iter) -> Result<T>
    |        --------- required by a bound in this function
302 | where
303 |     T: de::DeserializeOwned,
    |        ^^^^^^^^^^^^^^^^^^^^ required by this bound in `from_iter`
help: consider specifying the generic arguments
    |
25  |             envy::from_iter::<[(std::string::String, std::string::String); 1], T>([("INCLUDE_PATHS".into(), "linux,linux-old".into())]).unwrap(),
    |                            ++++++++++++++++++++++++++++++++++++++++++++++++++++++
```
while `assert_eq!(expected, actual)` works.

Reply:
That's odd how it doesn't automatically compute the types. You can use `PathBuf::from` there.

Using left as actual and right as expected is AFAIK common in Rust